syntax = "proto3";

package arduino_trader.trading;

import "common/common.proto";

// Trading Service - Handles trade execution
service TradingService {
  // Execute a single trade
  rpc ExecuteTrade(ExecuteTradeRequest) returns (ExecuteTradeResponse);

  // Execute multiple trades (batch)
  rpc BatchExecuteTrades(BatchExecuteTradesRequest) returns (BatchExecuteTradesResponse);

  // Get trade status
  rpc GetTradeStatus(GetTradeStatusRequest) returns (GetTradeStatusResponse);

  // Get trade history
  rpc GetTradeHistory(GetTradeHistoryRequest) returns (GetTradeHistoryResponse);

  // Cancel a pending trade
  rpc CancelTrade(CancelTradeRequest) returns (CancelTradeResponse);

  // Validate trade (pre-execution check)
  rpc ValidateTrade(ValidateTradeRequest) returns (ValidateTradeResponse);

  // Health check
  rpc HealthCheck(common.Empty) returns (common.HealthCheckResponse);
}

// Execute trade request
message ExecuteTradeRequest {
  string account_id = 1;
  string isin = 2;
  string symbol = 3;
  common.TradeSide side = 4;
  double quantity = 5;
  OrderType order_type = 6;
  common.Money limit_price = 7;  // Only for LIMIT orders
  common.Money stop_price = 8;   // Only for STOP orders
  TimeInForce time_in_force = 9;
  bool dry_run = 10;
}

// Order type
enum OrderType {
  ORDER_TYPE_UNSPECIFIED = 0;
  MARKET = 1;
  LIMIT = 2;
  STOP = 3;
  STOP_LIMIT = 4;
}

// Time in force
enum TimeInForce {
  TIME_IN_FORCE_UNSPECIFIED = 0;
  DAY = 1;
  GTC = 2;  // Good Till Cancelled
  IOC = 3;  // Immediate or Cancel
  FOK = 4;  // Fill or Kill
}

// Execute trade response
message ExecuteTradeResponse {
  bool success = 1;
  string trade_id = 2;
  string order_id = 3;
  common.TradeStatus status = 4;
  string message = 5;
  TradeExecution execution = 6;
}

// Trade execution details
message TradeExecution {
  string trade_id = 1;
  string order_id = 2;
  string isin = 3;
  string symbol = 4;
  common.TradeSide side = 5;
  double quantity_requested = 6;
  double quantity_filled = 7;
  common.Money average_price = 8;
  common.Money total_cost = 9;
  common.Money commission = 10;
  common.Money fees = 11;
  common.Timestamp executed_at = 12;
  string broker_reference = 13;
}

// Batch execute request
message BatchExecuteTradesRequest {
  string account_id = 1;
  repeated ExecuteTradeRequest trades = 2;
  bool stop_on_error = 3;
}

// Batch execute response
message BatchExecuteTradesResponse {
  bool all_success = 1;
  repeated ExecuteTradeResponse results = 2;
  int32 successful = 3;
  int32 failed = 4;
}

// Get trade status request
message GetTradeStatusRequest {
  string trade_id = 1;
}

// Get trade status response
message GetTradeStatusResponse {
  bool found = 1;
  TradeExecution execution = 2;
  common.TradeStatus status = 3;
}

// Get trade history request
message GetTradeHistoryRequest {
  string account_id = 1;
  int32 days = 2;
  string isin = 3;  // Optional filter
  common.TradeSide side = 4;  // Optional filter
  int32 limit = 5;
  int32 offset = 6;
}

// Get trade history response
message GetTradeHistoryResponse {
  repeated TradeExecution executions = 1;
  int32 total = 2;
}

// Cancel trade request
message CancelTradeRequest {
  string trade_id = 1;
  string reason = 2;
}

// Cancel trade response
message CancelTradeResponse {
  bool success = 1;
  string message = 2;
  common.TradeStatus new_status = 3;
}

// Validate trade request
message ValidateTradeRequest {
  string account_id = 1;
  string isin = 2;
  common.TradeSide side = 3;
  double quantity = 4;
  common.Money limit_price = 5;
}

// Validate trade response
message ValidateTradeResponse {
  bool valid = 1;
  repeated ValidationError errors = 2;
  repeated ValidationWarning warnings = 3;
  common.Money estimated_cost = 4;
  common.Money available_cash = 5;
}

// Validation error
message ValidationError {
  string code = 1;
  string message = 2;
  string field = 3;
}

// Validation warning
message ValidationWarning {
  string code = 1;
  string message = 2;
}
